pkg/querier/batch/stream.go

```diff
@@ -39,64 +45,120 @@ func (bs *batchStream) atTime() int64 {
+	return b.Timestamps[b.Index], b.SampleValues[b.Index]
+}
+
+func (bs *batchStream) atHistogram() (int64, *histogram.Histogram) {
+	b := &(*bs)[0]
+	return b.Timestamps[b.Index], b.HistogramValues[b.Index]
 }
 
+func (bs *batchStream) atFloatHistogram() (int64, *histogram.FloatHistogram) {
+	b := &(*bs)[0]
+	return b.Timestamps[b.Index], b.FloatHistogramValues[b.Index]
+}
+
+// mergeStreams merges streams of Batches of the same series over time.
+// Samples are simply merged by time when they are the same type (float/histogram/...), with the left stream taking precedence if the timestamps are equal.
+// When sample are different type, batches are not merged. In case of equal timestamps, histograms take precedence since they have more information.
 func mergeStreams(left, right batchStream, result batchStream, size int) batchStream {
+
 	// Reset the Index and Length of existing batches.
 	for i := range result {
 		result[i].Index = 0
 		result[i].Length = 0
 	}
+
 	resultLen := 1 // Number of batches in the final result.
 	b := &result[0]
 
-	// This function adds a new batch to the result
-	// if the current batch being appended is full.
-	checkForFullBatch := func() {
+	// Step to the next Batch in the result, create it if it does not exist
+	nextBatch := func(valueTypes chunkenc.ValueType) {
+		// The Index is the place at which new sample
+		// has to be appended, hence it tells the length.
+		b.Length = b.Index
+		resultLen++
+		if resultLen > len(result) {
+			// It is possible that result can grow longer
+			// then the one provided.
+			result = append(result, createBatch(valueTypes))
+		}
+		b = &result[resultLen-1]
+	}
+
+	ensureBatchType := func(valueTypes chunkenc.ValueType) {
+		for b.ValueTypes != valueTypes {
+			if b.ValueTypes == chunkenc.ValNone {
+				// Uninitialized Batch
+				initBatchType(b, valueTypes)
+				return
+			}
+			// These should be rare, means that we're merging into something where the chunk type changed
+			if b.Index == 0 {
+				// Batch not written yet
+				result[resultLen-1] = createBatch(valueTypes)
+				b = &result[resultLen-1]
+				return
+			}
+			// Batch already started, finish and start new
+			// We don't know its type so let the loop run one more time
+			nextBatch(valueTypes)
+		}
+	}
+
+	// Ensure that the batch we're writing to is not full and of the right type.
+	ensureBatch := func(valueTypes chunkenc.ValueType) {
 		if b.Index == size {
-			// The batch reached it intended size.
+			// The batch reached its intended size.
 			// Add another batch the the result
 			// and use it for further appending.
-
-			// The Index is the place at which new sample
-			// has to be appended, hence it tells the length.
-			b.Length = b.Index
-			resultLen++
-			if resultLen > len(result) {
-				// It is possible that result can grow longer
-				// then the one provided.
-				result = append(result, chunk.Batch{})
-			}
-			b = &result[resultLen-1]
+			nextBatch(valueTypes)
```

Comment:
Assuming the suggestion I left in `nextBatch()` is correct, then here you can return after calling `nextBatch()`. If we do return, then I would remove `ensureBatchType()` function at all and inline it below here, given it's just used here.

Reply:
reverted